pkg/loadbalancer/services.go

```diff
@@ -0,0 +1,415 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package loadbalancer
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"text/tabwriter"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/cilium/cilium/pkg/container"
+	"github.com/cilium/cilium/pkg/source"
+	"github.com/cilium/cilium/pkg/statedb"
+	"github.com/cilium/cilium/pkg/statedb/reconciler"
+)
+
+// Services provides validated write access to the service and backend tables.
+// This is currently experimental and not in use by default.
+// See comment in [tables.go] for further context.
+type Services struct {
+	db   *statedb.DB
+	svcs statedb.RWTable[*Service]
+	bes  statedb.RWTable[*ServiceBackend]
+
+	hooks []ServiceHook
+}
+
+func NewServices(cfg Config, db *statedb.DB, svcs statedb.RWTable[*Service], bes statedb.RWTable[*ServiceBackend]) (*Services, error) {
+	if !cfg.EnableNewServices {
+		return nil, nil
+	}
+	return &Services{db, svcs, bes, nil}, nil
+}
+
+func (s *Services) IsEnabled() bool {
+	return s != nil
+}
+
+// Hook is a function that runs when a service is upserted. It can perform reads using the provided
+// ReadTxn, but for consistency it must not depend on any other data (except when the data is constant).
+// The hook can modify the service object.
+type ServiceHook func(statedb.ReadTxn, *Service)
+
+type ServiceWriteTxn struct {
+	statedb.WriteTxn
+}
+
+var (
+	ErrServiceSourceMismatch = errors.New("service exists from different source")
+	ErrServiceConflict       = errors.New("conflict with a service with same address, but different name")
+)
+
+// RegisterHook adds a new hook to modify a service when upserted. This must only be called from
+// providers/invokes and not after starting!
+func (s *Services) RegisterHook(h ServiceHook) {
+	s.hooks = append(s.hooks, h)
+}
+
+// Services returns the service table for reading.
+// Convenience method for reducing dependencies for service writers.
+func (s *Services) Services() statedb.Table[*Service] {
+	return s.svcs
+}
+
+// Services returns the backend table for reading.
+// Convenience method for reducing dependencies for service writers.
+func (s *Services) Backends() statedb.Table[*ServiceBackend] {
+	return s.bes
+}
+
+// WriteTxn returns a write transaction against services & backends and other additional
+// tables to be used with the methods of [Services]. The returned transaction MUST be
+// Abort()'ed or Commit()'ed.
+func (s *Services) WriteTxn(extraTables ...statedb.TableMeta) ServiceWriteTxn {
+	return ServiceWriteTxn{
+		s.db.WriteTxn(s.svcs, append(extraTables, s.bes)...),
+	}
+}
+
+func (s *Services) UpsertService(txn ServiceWriteTxn, params *ServiceParams) (created bool, err error) {
+	var svc Service
+
+	existing, _, found := s.svcs.First(txn, ServiceL3n4AddrIndex.Query(params.Frontend))
+	if found {
+		// Do not allow services with same address but different names to override each other.
+		if !existing.Name.Equal(params.Name) {
+			return false, ErrServiceConflict
+		}
```

Comment:
Ah that's very interesting. I don't think Cilium currently handles such a possibility. But if it indeed is possible then yes we would need to have retries in the data source implementation.

Or an alternative is to be able to have overlapping services and allow this to happen in any order. E.g. `Shadowed []*Service` field to store the overlapping, but inactive services.

Reply:
I'd suggest asking ourselves what is a unique and immutable key for Frontends and then querying by this key. Does Frontend correspond to a pair (Service, Port) in Kubernetes? If so, then perhaps we should query by the fully qualified Service name (`namespace+name` in Kubernetes) plus the Port (including protocol).

Reply:
The unique immutable key is the tuple `(address, protocol, port)` (L3n4Addr). The service name is there to group multiple frontends together and to associate the backends with the frontends (it's a secondary index). 

We cannot have the service name in the key since then we may end up with multiple frontends with the same address and that's impossible to reconcile.